src/transformStateWithClones.js

```diff
@@ -8,6 +8,37 @@
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  let currentState = { ...state };
+  const states = [];
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+      case 'addProperties':
+        if (typeof action.extraData === 'object' && action.extraData !== null) {
+          Object.entries(action.extraData).forEach(([key, value]) => {
+            newState[key] = value;
+          });
+        }
+        break;
+
+      case 'removeProperties':
+        if (action.keysToRemove) {
+          for (const key of action.keysToRemove) {
+            delete newState[key];
+          }
+        }
```

Comment:
[CODE STYLE] - It's good that you are checking if `action.keysToRemove` exists before attempting to use it. However, you should also ensure that it's an array to avoid potential runtime errors if a different data type is passed.